From: Gunnar Hjalmarsson <gunnarhj@ubuntu.com>
Date: Tue, 7 Aug 2018 11:15:57 +0100
Subject: Show warning dialog in case of error in ~/.profile etc.

By using bash instead of sh, a syntax error in a sourced file does
not cause a fatal error, but the user is alerted via a dialog in the
same way as for other types of errors.
Bug-Ubuntu: https://launchpad.net/bugs/678421
Forwarded: https://bugzilla.gnome.org/738970
---
 data/Xsession.in            | 28 ++++++++++++++++++++++++----
 data/config-error-dialog.sh | 29 +++++++++++++++++++++++++++++
 data/meson.build            |  7 ++++++-
 po/POTFILES.in              |  1 +
 4 files changed, 60 insertions(+), 5 deletions(-)
 create mode 100644 data/config-error-dialog.sh

diff --git a/data/Xsession.in b/data/Xsession.in
index 40da9c7..54b1f9a 100755
--- a/data/Xsession.in
+++ b/data/Xsession.in
@@ -29,12 +29,32 @@ command="$@"
 # good for debugging where things went wrong
 echo "$0: Beginning session setup..."
 
+# temporary storage of error messages
+ERR=$(mktemp --tmpdir config-err-XXXXXX)
+
+source_with_error_check () {
+  CONFIG_FILE="$1"
+  . "$CONFIG_FILE" 2>"$ERR"
+  if [ -s "$ERR" ]; then
+    . /etc/gdm3/config-error-dialog.sh
+  fi
+  cat "$ERR" >&2
+  truncate -s 0 "$ERR"
+}
+
 # First read /etc/profile and .profile
-test -f /etc/profile && . /etc/profile
-test -f "$HOME/.profile" && . "$HOME/.profile"
+for file in /etc/profile "$HOME/.profile"; do
+  if [ -f "$file" ]; then
+    source_with_error_check "$file"
+  fi
+done
+
 # Second read /etc/xprofile and .xprofile for X specific setup
-test -f /etc/xprofile && . /etc/xprofile
-test -f "$HOME/.xprofile" && . "$HOME/.xprofile"
+for file in /etc/xprofile "$HOME/.xprofile"; do
+  if [ -f "$file" ]; then
+    source_with_error_check "$file"
+  fi
+done
 
 # Translation stuff
 if [ -x "@libexecdir@/gdmtranslate" ] ; then
diff --git a/data/config-error-dialog.sh b/data/config-error-dialog.sh
new file mode 100644
index 0000000..08cb7b9
--- /dev/null
+++ b/data/config-error-dialog.sh
@@ -0,0 +1,29 @@
+# Copyright (C) 2014 Canonical Ltd
+# Author: Gunnar Hjalmarsson <gunnarhj@ubuntu.com>
+#
+# This program is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the Free Software
+# Foundation, version 3 of the License.
+#
+# See http://www.gnu.org/copyleft/gpl.html the full text of the license.
+
+# This file may be sourced by the function source_with_error_check() in
+# /etc/gdm/Xsession
+
+export TEXTDOMAIN=gdm
+. /usr/bin/gettext.sh
+
+PARA1=$(eval_gettext 'Error found when loading $CONFIG_FILE:')
+PARA2=$(gettext 'As a result the session will not be configured correctly.
+You should fix the problem as soon as feasible.')
+
+TEXT="$PARA1\n\n$(fold -s $ERR)\n\n$PARA2"
+
+if [ -x /usr/bin/kdialog ]; then
+	TEXT_FILE=$(mktemp --tmpdir config-err-kdialog-XXXXXX)
+	echo -n "$TEXT" > "$TEXT_FILE"
+	kdialog --textbox "$TEXT_FILE" 500 300
+	rm -f "$TEXT_FILE"
+elif [ -x /usr/bin/zenity ]; then
+	zenity --warning --no-wrap --text="$TEXT"
+fi
diff --git a/data/meson.build b/data/meson.build
index 9ae2d4b..ddce3fe 100644
--- a/data/meson.build
+++ b/data/meson.build
@@ -178,8 +178,13 @@ if get_option('gdm-xsession')
     output: '@BASENAME@',
     configuration: {
       'libexecdir': gdm_prefix / get_option('libexecdir'),
-      'XSESSION_SHELL': get_option('solaris')? '/bin/ksh' : '/bin/sh',
+      'XSESSION_SHELL': get_option('solaris')? '/bin/ksh' : '/bin/bash',
     },
     install_dir: gdmconfdir,
   )
+
+  install_data(
+    'config-error-dialog.sh',
+    install_dir: gdmconfdir,
+  )
 endif
diff --git a/po/POTFILES.in b/po/POTFILES.in
index e84eef2..0486c56 100644
--- a/po/POTFILES.in
+++ b/po/POTFILES.in
@@ -38,6 +38,7 @@ daemon/gdm-x-session.c
 daemon/main.c
 daemon/session-worker-main.c
 daemon/test-session-client.c
+data/config-error-dialog.sh
 data/org.gnome.login-screen.gschema.xml
 libgdm/gdm-user-switching.c
 utils/gdmflexiserver.c
